util/encoding.go

```diff
@@ -0,0 +1,243 @@
+	bs.appendNBits([]byte{b}, 8)
+}
+
+// WriteUInt12 writes the rightmost 12 bits in b into the BitStream.
+// For instance, the input b is 0xff01, the effective part refers to the rightmost 12 bits,
+// which should be 0xf01.
+func (bs *BitStream) WriteUInt12(b uint16) {
+	first := byte(b >> 4)
+	second := byte(b << 4)
+	bs.appendNBits([]byte{first, second}, 12)
+}
+
+// WriteUInt16 writes a full uint16 (two bytes) into the BitStream.
+func (bs *BitStream) WriteUInt16(b uint16) {
+	first := byte(b >> 8)
+	second := byte(b)
+	bs.appendNBits([]byte{first, second}, 16)
+}
+
+// WriteTwoBitField encapsulates WriteByte2 to get convenience to some extent, when encoding GPP strings.
+func (bs *BitStream) WriteTwoBitField(bList []byte) {
+	for _, b := range bList {
+		bs.WriteByte2(b)
+	}
+}
+
+/*
+	WriteFibonacciInt writes int based on Fibonacci Coding https://en.wikipedia.org/wiki/Fibonacci_coding.
+
+By definition, Fibonacci numbers are numbers that are the sum of the two Fibonacci numbers that come before.
+The numbers are [excluding 0, 1,] 1 (0+1), 2 (1+1), 3 (1+2), 5 (2+3), 8 (3+5), etc. All of the integers that
+are larger and equal to 0 are able to split into a combination of at least one Fibonacci numbers. For instance,
+6 refers to 1+5, while 7 refers to 2+5, where 1, 2, 5 are Fibonacci numbers.
+
+To encode an integer using Fibonacci Coding, just follow these steps:
+1.Find the largest Fibonacci number equal to or less than N; subtract this number from N, keeping track of the remainder.
+2.If the number subtracted was the i-th Fibonacci number F(i), put a 1 in place i−2 in the code word (counting the left-most digit as place 0).
+3.Repeat the previous steps, substituting the remainder for N, until a remainder of 0 is reached.
+4.Place an additional 1 after the rightmost digit in the code word.
+
+So a Fibonacci encoded bit sequence is all about marking the corresponding bit as 1 while leaving the others as 0.
+Here are some examples:
+
+	FibonacciInt(1) = 11			1-> (1*1)->1
+	FibonacciInt(2) = 011 			2-> (1*1+2*1)->01
+	FibonacciInt(3) = 0011			3-> (1*0+2*0+3*1)->001
+	FibonacciInt(4) = 1011			4-> (1*1+2*0+3*1)->101
+	FibonacciInt(12) = 101011		12->(1*1+2*0+3*1+5*0+8*1)->10101
+
+According to the algorithm, no sequential ones can be found in the bit sequence, and the final bit is always a single 1,
+which makes it easy to extract several continuous Fibonacci encoded integers out from a bit stream if we
+place an additional 1 in the end, since 11 could be an end-of-sequence indicator. The advantage of this is,
+that a sequence of numbers can be encoded into a sequence of bits without the need to know the length of the
+bit sequences in advance.
+*/
+func (bs *BitStream) WriteFibonacciInt(num uint16) error {
+	// The num should be [1,6765). Actually once the num is larger than or equal to 987,
+	// the efficiency of Fibonacci Encoding would be no better than 'WriteUint16'.
+	if num <= 0 || num >= fibonacci(fibLen) {
```

Comment:
If you are limiting to `fibLen`, then you can just lookup in the `fibLookup[i]` array rather than calling the `fibonacci()` function. The function just adds code to handle the case where the number if greater than `fibLen`.